# Capstone Project Week 3:
# Part 3  
## Segmenting and Clustering Neighborhoods in Toronto: 3rd task 

### 3rd Task: Utilize the Foursquare location data and, Explore and Cluster the Neighborhoods in Toronto.

For this assignment, I am required to explore and cluster the neighborhoods in Toronto.
I have separated the assignment into 3 parts.

Part 1:  1st task : Download the data set, Scraple and wrangle:
                    Clean the data set:
                    Merge raws with neighborhoods that belong to the same Borough. 
                    I am using the pandas library to scrap the Data.
        
Part 2:             This the same as part 1 with a
        2nd task :  Find the latitude and longitude of each neiborhood.

Part 3: In the 3rd Notebook I will use the beautifull soup package ( for practising) and perform 1st, 2nd and, 3rd task. 
        3rd task : Explore and cluster the Neiborhoods in Toronto.


## Part 3 :Table of Contents



<font size = 3>

3.1. <a href="#item1">Explore Neighborhoods in Toronto </a>

3.1.1. <a href="#item2">Visualise the Boroughs of Toronto that have in their names the word Toronto.</a>

3.1.2. <a href="#item3">Utilizing the Foursquare API to explore the neighborhoods and segment them.</a>

3.1.3. <a href="#item4"> Explore only one neighborhood from our dataframe.</a>

3.1.2. <a href="#item5"> Explore many Neighborhoods in Boroughs with the name Toronto</a>    





<font size = 3>

3.2. <a href="#item1"> Analysis of each Neighborhood  </a>

3.3. <a href="#item2"> Cluster of the Neighborhoods with K-means clustering algorithm </a>

3.3.1. <a href="#item3"> Run K-means to cluster the neighborhoods into 5 clusters and visualise them</a>

3.3.2  <a href="#item4"> Visualise the clustering at neighbourhoods of the Boroughs that contain the name Toronto</a>

3.3.4. <a href="#item5">  Examine Clusters</a>    
</font>
</div>

## Part 1

## Scrap the wickipedia page using BeautifulSoup, to obtain the data of postcodes of Toronto.

I need to scrape the following Wikipedia page, https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M, in order to obtain the data that is in the table of postal codes and to transform the data into a pandas dataframe

Before I get the data and start exploring it, let's download all the dependencies that I will need. For Part 3 (which is the complete version of the project) I will use all the following libraries for analysing my data.

In [ ]:
import numpy as np # library to handle data in a vectorized manner
import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values
import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# ##Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# ##import k-means from clustering stage
from sklearn.cluster import KMeans
from sklearn.datasets.samples_generator import make_blobs
!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library
from bs4 import BeautifulSoup
import lxml

print('Libraries imported.')

### Download Data sets with Beautifull soup, scrape and wrangle

In this Notebook I am using the BeautifulSoup library to scrap the data for practising reasons. At the previous notebooks (Part 1,Part 2) I have used pandas.

In [10]:
# download data and parse it:
wiki = requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M').text # add.text
soup = BeautifulSoup(wiki,'lxml') # Create an object soup 
print(soup.title) # The title is printed to see if the page has been scraped succesfully

<title>List of postal codes of Canada: M - Wikipedia</title>


In [11]:
#  I want to display the image of the table therefore I import IPython library
from IPython.display import display_html # Library for images

tab = str(soup.table)
display_html(tab,raw=True)

Postal Code
 
 Borough
 
 Neighborhood
 
 
 M1A
 
 Not assigned
 
 Not assigned
 
 
 M2A
 
 Not assigned
 
 Not assigned
 
 
 M3A
 
 North York
 
 Parkwoods
 
 
 M4A
 
 North York
 
 Victoria Village
 
 
 M5A
 
 Downtown Toronto
 
 Regent Park, Harbourfront
 
 
 M6A
 
 North York
 
 Lawrence Manor, Lawrence Heights
 
 
 M7A
 
 Downtown Toronto
 
 Queen's Park, Ontario Provincial Government
 
 
 M8A
 
 Not assigned
 
 Not assigned
 
 
 M9A
 
 Etobicoke
 
 Islington Avenue, Humber Valley Village
 
 
 M1B
 
 Scarborough
 
 Malvern, Rouge
 
 
 M2B
 
 Not assigned
 
 Not assigned
 
 
 M3B
 
 North York
 
 Don Mills
 
 
 M4B
 
 East York
 
 Parkview Hill, Woodbine Gardens
 
 
 M5B
 
 Downtown Toronto
 
 Garden District, Ryerson
 
 
 M6B
 
 North York
 
 Glencairn
 
 
 M7B
 
 Not assigned
 
 Not assigned
 
 
 M8B
 
 Not assigned
 
 Not assigned
 
 
 M9B
 
 Etobicoke
 
 West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
 
 
 M1C
 
 Scarborough
 
 Rouge Hill, Port Union, Highland Creek
 
 
 M2C
 
 Not assigned
 
 Not assigned
 
 
 M3C
 
 North York
 
 Don Mills
 
 
 M4C
 
 East York
 
 Woodbine Heights
 
 
 M5C
 
 Downtown Toronto
 
 St. James Town
 
 
 M6C
 
 York
 
 Humewood-Cedarvale
 
 
 M7C
 
 Not assigned
 
 Not assigned
 
 
 M8C
 
 Not assigned
 
 Not assigned
 
 
 M9C
 
 Etobicoke
 
 Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
 
 
 M1E
 
 Scarborough
 
 Guildwood, Morningside, West Hill
 
 
 M2E
 
 Not assigned
 
 Not assigned
 
 
 M3E
 
 Not assigned
 
 Not assigned
 
 
 M4E
 
 East Toronto
 
 The Beaches
 
 
 M5E
 
 Downtown Toronto
 
 Berczy Park
 
 
 M6E
 
 York
 
 Caledonia-Fairbanks
 
 
 M7E
 
 Not assigned
 
 Not assigned
 
 
 M8E
 
 Not assigned
 
 Not assigned
 
 
 M9E
 
 Not assigned
 
 Not assigned
 
 
 M1G
 
 Scarborough
 
 Woburn
 
 
 M2G
 
 Not assigned
 
 Not assigned
 
 
 M3G
 
 Not assigned
 
 Not assigned
 
 
 M4G
 
 East York
 
 Leaside
 
 
 M5G
 
 Downtown Toronto
 
 Central Bay Street
 
 
 M6G
 
 Downtown Toronto
 
 Christie
 
 
 M7G
 
 Not assigned
 
 Not assigned
 
 
 M8G
 
 Not assigned
 
 Not assigned
 
 
 M9G
 
 Not assigned
 
 Not assigned
 
 
 M1H
 
 Scarborough
 
 Cedarbrae
 
 
 M2H
 
 North York
 
 Hillcrest Village
 
 
 M3H
 
 North York
 
 Bathurst Manor, Wilson Heights, Downsview North
 
 
 M4H
 
 East York
 
 Thorncliffe Park
 
 
 M5H
 
 Downtown Toronto
 
 Richmond, Adelaide, King
 
 
 M6H
 
 West Toronto
 
 Dufferin, Dovercourt Village
 
 
 M7H
 
 Not assigned
 
 Not assigned
 
 
 M8H
 
 Not assigned
 
 Not assigned
 
 
 M9H
 
 Not assigned
 
 Not assigned
 
 
 M1J
 
 Scarborough
 
 Scarborough Village
 
 
 M2J
 
 North York
 
 Fairview, Henry Farm, Oriole
 
 
 M3J
 
 North York
 
 Northwood Park, York University
 
 
 M4J
 
 East York
 
 East Toronto, Broadview North (Old East York)
 
 
 M5J
 
 Downtown Toronto
 
 Harbourfront East, Union Station, Toronto Islands
 
 
 M6J
 
 West Toronto
 
 Little Portugal, Trinity
 
 
 M7J
 
 Not assigned
 
 Not assigned
 
 
 M8J
 
 Not assigned
 
 Not assigned
 
 
 M9J
 
 Not assigned
 
 Not assigned
 
 
 M1K
 
 Scarborough
 
 Kennedy Park, Ionview, East Birchmount Park
 
 
 M2K
 
 North York
 
 Bayview Village
 
 
 M3K
 
 North York
 
 Downsview
 
 
 M4K
 
 East Toronto
 
 The Danforth West, Riverdale
 
 
 M5K
 
 Downtown Toronto
 
 Toronto Dominion Centre, Design Exchange
 
 
 M6K
 
 West Toronto
 
 Brockton, Parkdale Village, Exhibition Place
 
 
 M7K
 
 Not assigned
 
 Not assigned
 
 
 M8K
 
 Not assigned
 
 Not assigned
 
 
 M9K
 
 Not assigned
 
 Not assigned
 
 
 M1L
 
 Scarborough
 
 Golden Mile, Clairlea, Oakridge
 
 
 M2L
 
 North York
 
 York Mills, Silver Hills
 
 
 M3L
 
 North York
 
 Downsview
 
 
 M4L
 
 East Toronto
 
 India Bazaar, The Beaches West
 
 
 M5L
 
 Downtown Toronto
 
 Commerce Court, Victoria Hotel
 
 
 M6L
 
 North York
 
 North Park, Maple Leaf Park, Upwood Park
 
 
 M7L
 
 Not assigned
 
 Not assigned
 
 
 M8L
 
 Not assigned
 
 Not assigned
 
 
 M9L
 
 North York
 
 Humber Summit
 
 
 

The html table needs to be be converted to pandas dataFreme for processing 

### Create DataFrame in Pandas with PostalCode, Borough, and Neighborhood  and Clean the Data

In [12]:
df = pd.read_html(tab)
df_raw=df[0]
df_raw.head()

Postal Code           Borough               Neighborhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

#### Clean the Data set: The DataFrame has NA inputs that I need to remove them 

In [13]:
df_new = df_raw[df_raw.Borough != 'Not assigned'] # clean the df_raw from the cells that is 'Not assigned' and name df_new

df_new.head(15)

Postal Code           Borough  \
2          M3A        North York   
3          M4A        North York   
4          M5A  Downtown Toronto   
5          M6A        North York   
6          M7A  Downtown Toronto   
8          M9A         Etobicoke   
9          M1B       Scarborough   
11         M3B        North York   
12         M4B         East York   
13         M5B  Downtown Toronto   
14         M6B        North York   
17         M9B         Etobicoke   
18         M1C       Scarborough   
20         M3C        North York   
21         M4C         East York   

                                         Neighborhood  
2                                           Parkwoods  
3                                    Victoria Village  
4                           Regent Park, Harbourfront  
5                    Lawrence Manor, Lawrence Heights  
6         Queen's Park, Ontario Provincial Government  
8             Islington Avenue, Humber Valley Village  
9                                      Malvern, Rouge  
11                                          Don Mills  
12                    Parkview Hill, Woodbine Gardens  
13                           Garden District, Ryerson  
14                                          Glencairn  
17  West Deane Park, Princess Gardens, Martin Grov...  
18             Rouge Hill, Port Union, Highland Creek  
20                                          Don Mills  
21                                   Woodbine Heights

According to the data, there are 8 Borought in Toronto with neighborhoods

In [6]:
df_new.loc[df_new.Neighborhood == 'Not assigned']# Find if there is any 'not assigned', here is totaly cleaned **************

Empty DataFrame
Columns: [Postal Code, Borough, Neighborhood]
Index: []

If they would be Neighborhood with 'not assigned', then I would make the neighborhood the same as the borough
I would have used the following script:



In [73]:
## df_new.Neighborhood.replace('Not assigned',df_new.Borough,inplace=True)##

More than one neighborhood can exist in one postal code area. For example, in the table on the Wikipedia page, you will notice that M3A, M4A, M5A are listed twice and has two neighborhoods: M5A (Harbourfront and Regent Park). These two rows will be combined into one row with the neighborhoods separated with a comma as shown in row 11 in the above table.

In [14]:
df_toronto = df_new.groupby(['Postal Code', 'Borough'])['Neighborhood'].apply(lambda x: ', '.join(x)) #Here I connect all the neiborhoods with the same postcodes and separate them with a comma
df_toronto = df_toronto.reset_index()

In [15]:
df_toronto.head(20)

Postal Code      Borough                                       Neighborhood
0          M1B  Scarborough                                     Malvern, Rouge
1          M1C  Scarborough             Rouge Hill, Port Union, Highland Creek
2          M1E  Scarborough                  Guildwood, Morningside, West Hill
3          M1G  Scarborough                                             Woburn
4          M1H  Scarborough                                          Cedarbrae
5          M1J  Scarborough                                Scarborough Village
6          M1K  Scarborough        Kennedy Park, Ionview, East Birchmount Park
7          M1L  Scarborough                    Golden Mile, Clairlea, Oakridge
8          M1M  Scarborough    Cliffside, Cliffcrest, Scarborough Village West
9          M1N  Scarborough                        Birch Cliff, Cliffside West
10         M1P  Scarborough  Dorset Park, Wexford Heights, Scarborough Town...
11         M1R  Scarborough                                  Wexford, Maryvale
12         M1S  Scarborough                                          Agincourt
13         M1T  Scarborough            Clarks Corners, Tam O'Shanter, Sullivan
14         M1V  Scarborough  Milliken, Agincourt North, Steeles East, L'Amo...
15         M1W  Scarborough                      Steeles West, L'Amoreaux West
16         M1X  Scarborough                                        Upper Rouge
17         M2H   North York                                  Hillcrest Village
18         M2J   North York                       Fairview, Henry Farm, Oriole
19         M2K   North York                                    Bayview Village

In [16]:
df_toronto.shape

(103, 3)

## Part 2

## Find the latitude and longitude coordinates of each neighborhood

According to the instructions I will try to use the geocoder to get the lattitude and longitude, if that fails I will import a file with the longitute and latitude.

In [16]:
#import geocoder # import geocoder

# initialize your variable to None
#lat_lng_coords = None

# loop until you get the coordinates
#while(lat_lng_coords is None):
  #g = geocoder.google('{}, Toronto, Ontario'.format(postal_code))
#  lat_lng_coords = g.latlng

#latitude = lat_lng_coords[0]
#longitude = lat_lng_coords[1]

The previous code was not executed succesfull and as the course leaders gave the following alternative, I am using it here: http://cocl.us/Geospatial_data to download in the server as csv file

In [17]:
#add Geo-spatial data
df_toronto_ll= pd.read_csv("http://cocl.us/Geospatial_data")

#-----Merge the two data sets in one----
Toronto_data=pd.merge(df_toronto, df_toronto_ll) 
Toronto_data.head(10)


Postal Code      Borough                                     Neighborhood  \
0         M1B  Scarborough                                   Malvern, Rouge   
1         M1C  Scarborough           Rouge Hill, Port Union, Highland Creek   
2         M1E  Scarborough                Guildwood, Morningside, West Hill   
3         M1G  Scarborough                                           Woburn   
4         M1H  Scarborough                                        Cedarbrae   
5         M1J  Scarborough                              Scarborough Village   
6         M1K  Scarborough      Kennedy Park, Ionview, East Birchmount Park   
7         M1L  Scarborough                  Golden Mile, Clairlea, Oakridge   
8         M1M  Scarborough  Cliffside, Cliffcrest, Scarborough Village West   
9         M1N  Scarborough                      Birch Cliff, Cliffside West   

    Latitude  Longitude  
0  43.806686 -79.194353  
1  43.784535 -79.160497  
2  43.763573 -79.188711  
3  43.770992 -79.216917  
4  43.773136 -79.239476  
5  43.744734 -79.239476  
6  43.727929 -79.262029  
7  43.711112 -79.284577  
8  43.716316 -79.239476  
9  43.692657 -79.264848

For the examiner: In order to crossexamine the results with the table at the instructions I've printed the first raw of the table at the instructions.

In [18]:
Toronto_data.loc[Toronto_data.Neighborhood == 'Central Bay Street'] #  

Postal Code           Borough        Neighborhood   Latitude  Longitude
57         M5G  Downtown Toronto  Central Bay Street  43.657952 -79.387383

In [19]:
Toronto_data.shape

(103, 5)

Toronto_data, with all the neighbourhoods of Toronto has 103 rows. I need to focus the analysis in a smaller data set of boroughs

# Part 3

 This part includes the  Clustering and Exploration of the neighbourhoods of Toronto which contain Toronto in the name of their Borough.


# 3.1 Explore Neighborhoods in Toronto 

### Getting all the rows of the Toronto_data that contains the word Toronto

In [20]:
Toronto_T = Toronto_data[Toronto_data['Borough'].str.contains('Toronto',regex=False)] 
#Toronto_T = Toronto_data[Toronto_data['Borough'] =='Toronto'].reset_index(drop=True) :::It doesnt work as there aren't Boroughts with just Toronto tha
Toronto_T # These are the boroughts of toronto that have the word Toronto in their names 

Postal Code           Borough  \
37         M4E      East Toronto   
41         M4K      East Toronto   
42         M4L      East Toronto   
43         M4M      East Toronto   
44         M4N   Central Toronto   
45         M4P   Central Toronto   
46         M4R   Central Toronto   
47         M4S   Central Toronto   
48         M4T   Central Toronto   
49         M4V   Central Toronto   
50         M4W  Downtown Toronto   
51         M4X  Downtown Toronto   
52         M4Y  Downtown Toronto   
53         M5A  Downtown Toronto   
54         M5B  Downtown Toronto   
55         M5C  Downtown Toronto   
56         M5E  Downtown Toronto   
57         M5G  Downtown Toronto   
58         M5H  Downtown Toronto   
59         M5J  Downtown Toronto   
60         M5K  Downtown Toronto   
61         M5L  Downtown Toronto   
63         M5N   Central Toronto   
64         M5P   Central Toronto   
65         M5R   Central Toronto   
66         M5S  Downtown Toronto   
67         M5T  Downtown Toronto   
68         M5V  Downtown Toronto   
69         M5W  Downtown Toronto   
70         M5X  Downtown Toronto   
75         M6G  Downtown Toronto   
76         M6H      West Toronto   
77         M6J      West Toronto   
78         M6K      West Toronto   
82         M6P      West Toronto   
83         M6R      West Toronto   
84         M6S      West Toronto   
85         M7A  Downtown Toronto   
87         M7Y      East Toronto   

                                         Neighborhood   Latitude  Longitude  
37                                        The Beaches  43.676357 -79.293031  
41                       The Danforth West, Riverdale  43.679557 -79.352188  
42                     India Bazaar, The Beaches West  43.668999 -79.315572  
43                                    Studio District  43.659526 -79.340923  
44                                      Lawrence Park  43.728020 -79.388790  
45                                   Davisville North  43.712751 -79.390197  
46                  North Toronto West, Lawrence Park  43.715383 -79.405678  
47                                         Davisville  43.704324 -79.388790  
48                        Moore Park, Summerhill East  43.689574 -79.383160  
49  Summerhill West, Rathnelly, South Hill, Forest...  43.686412 -79.400049  
50                                           Rosedale  43.679563 -79.377529  
51                        St. James Town, Cabbagetown  43.667967 -79.367675  
52                               Church and Wellesley  43.665860 -79.383160  
53                          Regent Park, Harbourfront  43.654260 -79.360636  
54                           Garden District, Ryerson  43.657162 -79.378937  
55                                     St. James Town  43.651494 -79.375418  
56                                        Berczy Park  43.644771 -79.373306  
57                                 Central Bay Street  43.657952 -79.387383  
58                           Richmond, Adelaide, King  43.650571 -79.384568  
59  Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752  
60           Toronto Dominion Centre, Design Exchange  43.647177 -79.381576  
61                     Commerce Court, Victoria Hotel  43.648198 -79.379817  
63                                           Roselawn  43.711695 -79.416936  
64    Forest Hill North & West, Forest Hill Road Park  43.696948 -79.411307  
65                The Annex, North Midtown, Yorkville  43.672710 -79.405678  
66                     University of Toronto, Harbord  43.662696 -79.400049  
67          Kensington Market, Chinatown, Grange Park  43.653206 -79.400049  
68  CN Tower, King and Spadina, Railway Lands, Har...  43.628947 -79.394420  
69                                     Stn A PO Boxes  43.646435 -79.374846  
70             First Canadian Place, Underground city  43.648429 -79.382280  
75                                           Christie  43.669542 -79.422564  
76                       Dufferin, Dovercourt Village  43.669005 -

I observe that the indexes  do not start from 0 and they are not continius numbers therefore I need to reset their indexes in order to analyse them.

In [26]:
Toronto_T = Toronto_T.reset_index(drop=True) # I reset the indexes to start from 0
Toronto_T.head(10)

Postal Code          Borough  \
0         M4E     East Toronto   
1         M4K     East Toronto   
2         M4L     East Toronto   
3         M4M     East Toronto   
4         M4N  Central Toronto   
5         M4P  Central Toronto   
6         M4R  Central Toronto   
7         M4S  Central Toronto   
8         M4T  Central Toronto   
9         M4V  Central Toronto   

                                        Neighborhood   Latitude  Longitude  
0                                        The Beaches  43.676357 -79.293031  
1                       The Danforth West, Riverdale  43.679557 -79.352188  
2                     India Bazaar, The Beaches West  43.668999 -79.315572  
3                                    Studio District  43.659526 -79.340923  
4                                      Lawrence Park  43.728020 -79.388790  
5                                   Davisville North  43.712751 -79.390197  
6                  North Toronto West, Lawrence Park  43.715383 -79.405678  
7                                         Davisville  43.704324 -79.388790  
8                        Moore Park, Summerhill East  43.689574 -79.383160  
9  Summerhill West, Rathnelly, South Hill, Forest...  43.686412 -79.400049

In [22]:
Toronto_T.shape

(39, 5)

## 3.1.2 Visualise  the Boroughs of Toronto that have in their names the word Toronto.
To get the geografical coordinates of Toronto I use the goepy library

In order to define an instance of the geocoder, we need to define a user_agent. We will name our agent toronto_explorer, as shown below

In [23]:
address = 'Toronto'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


#### Note for the reviewer: I am not sure if you could see the map in the GitHub. If you run the code the map appears

In [27]:
map_toronto = folium.Map(location=[43.6534817,-79.3839347],zoom_start=10)

for lat,lng,borough,neighborhood in zip(Toronto_T['Latitude'],Toronto_T['Longitude'],Toronto_T['Borough'],Toronto_T['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
    [lat,lng],
    radius=5,
    popup=label,
    color='magenta',
    fill=True,
    fill_color='#3186cc',
    fill_opacity=0.7,
    parse_html=False).add_to(map_toronto)
map_toronto

## 3.1.3 Utilizing the Foursquare API to explore the neighborhoods and segment them.

 ### Define Foursquare Credentials and version

In [28]:

CLIENT_ID = '04KBULDSADSDGUNFH5EBBFTRYS3AZRALGMQWCEUBCEMMZ2ZB' # your Foursquare ID
CLIENT_SECRET = 'Z2GVW1AYYWS5NXX3Z1BXRMZWJ4ZTLL3PHBETTPGWNYLMRAYW' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)


Your credentails:
CLIENT_ID: 04KBULDSADSDGUNFH5EBBFTRYS3AZRALGMQWCEUBCEMMZ2ZB
CLIENT_SECRET:Z2GVW1AYYWS5NXX3Z1BXRMZWJ4ZTLL3PHBETTPGWNYLMRAYW


## 3.1.4: Explore only one neighborhood from  our dataframe.

In [29]:
## Get the neighborhoods name
Toronto_T.loc[2,'Neighborhood'] # The third neighborhood in Toronto_T data has the index 2

'India Bazaar, The Beaches West'

Get the neighborhood's latitude and longitude values.

In [30]:
neighborhood_latitude = Toronto_T.loc[2, 'Latitude'] #  3rd neighborhood latitude value
neighborhood_longitude = Toronto_T.loc[2, 'Longitude'] # 3rd neighborhood longitude value

neighborhood_name = Toronto_T.loc[2, 'Neighborhood'] # 3rd neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of India Bazaar, The Beaches West are 43.6689985, -79.31557159999998.


Lets exprlore the top 100 venues that are in India Bazaar, The Beaches Wes within a radious of 500 meters 

First, let's create the GET request URL. I name  URL url.

In [31]:
LIMIT = 100 # limit of number of venues returned by Foursquare API


radius = 500 # define radius


 # create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url # display URL

'https://api.foursquare.com/v2/venues/explore?&client_id=04KBULDSADSDGUNFH5EBBFTRYS3AZRALGMQWCEUBCEMMZ2ZB&client_secret=Z2GVW1AYYWS5NXX3Z1BXRMZWJ4ZTLL3PHBETTPGWNYLMRAYW&v=20180605&ll=43.6689985,-79.31557159999998&radius=500&limit=100'

Send the get request and examine the results

In [32]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5eff588ce116f90c6c147e2d'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'East End',
  'headerFullLocation': 'East End, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 23,
  'suggestedBounds': {'ne': {'lat': 43.6734985045, 'lng': -79.30936207594885},
   'sw': {'lat': 43.6644984955, 'lng': -79.32178112405111}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4ade390ff964a5200e7421e3',
       'name': 'System Fitness',
       'location': {'address': '1661 Queen St East',
        'crossStreet': 'at Kingston Rd',
        'lat': 43.667171452103204,
        'lng': -79.31273345707406,
        'labeledLatLngs': [{'label': 'di

From the Foursquare API we know that all the information is in the *items* key. Let's borrow the *get_category_type* function to extract the category of the venue.

In [33]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

Now lets clean the json and structure it into a pandas dataframe

In [34]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head(10)

name            categories        lat        lng
0              System Fitness                   Gym  43.667171 -79.312733
1  British Style Fish & Chips     Fish & Chips Shop  43.668723 -79.317139
2         The Burger's Priest  Fast Food Restaurant  43.666731 -79.315556
3           Brett's Ice Cream        Ice Cream Shop  43.667222 -79.312831
4                     O Sushi      Sushi Restaurant  43.666684 -79.316614
5               Woodbine Park                  Park  43.664860 -79.315109
6            Godspeed Brewery               Brewery  43.672620 -79.319228
7                Murphy's Law                   Pub  43.667319 -79.312656
8                        LCBO          Liquor Store  43.666732 -79.314966
9             Casa di Giorgio    Italian Restaurant  43.666645 -79.315204

In [35]:
print('{} venues were returned by Foursquare at the India Bazaar, The Beaches West neighborhood.'.format(nearby_venues.shape[0]))

23 venues were returned by Foursquare at the India Bazaar, The Beaches West neighborhood.


## 3.1.5 Explore many Neighborhoods in Boroughs with the name Toronto

Creat a function to obtain the venues from all the neighborhoods in the boroughts that have Toronto in data frame Toronto_T

In [36]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [38]:
toronto_venues = getNearbyVenues(names=Toronto_T['Neighborhood'],
                                   latitudes=Toronto_T['Latitude'],
                                   longitudes=Toronto_T['Longitude']
                                  )

The Beaches
The Danforth West, Riverdale
India Bazaar, The Beaches West
Studio District
Lawrence Park
Davisville North
North Toronto West, Lawrence Park
Davisville
Moore Park, Summerhill East
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
Rosedale
St. James Town, Cabbagetown
Church and Wellesley
Regent Park, Harbourfront
Garden District, Ryerson
St. James Town
Berczy Park
Central Bay Street
Richmond, Adelaide, King
Harbourfront East, Union Station, Toronto Islands
Toronto Dominion Centre, Design Exchange
Commerce Court, Victoria Hotel
Roselawn
Forest Hill North & West, Forest Hill Road Park
The Annex, North Midtown, Yorkville
University of Toronto, Harbord
Kensington Market, Chinatown, Grange Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
Stn A PO Boxes
First Canadian Place, Underground city
Christie
Dufferin, Dovercourt Village
Little Portugal, Trinity
Brockton, Parkdale Village, Exhibition Place
High 

In [40]:
print(toronto_venues.shape)
toronto_venues.head(15)

(1614, 7)


Neighborhood  Neighborhood Latitude  \
0                    The Beaches              43.676357   
1                    The Beaches              43.676357   
2                    The Beaches              43.676357   
3                    The Beaches              43.676357   
4                    The Beaches              43.676357   
5   The Danforth West, Riverdale              43.679557   
6   The Danforth West, Riverdale              43.679557   
7   The Danforth West, Riverdale              43.679557   
8   The Danforth West, Riverdale              43.679557   
9   The Danforth West, Riverdale              43.679557   
10  The Danforth West, Riverdale              43.679557   
11  The Danforth West, Riverdale              43.679557   
12  The Danforth West, Riverdale              43.679557   
13  The Danforth West, Riverdale              43.679557   
14  The Danforth West, Riverdale              43.679557   

    Neighborhood Longitude                               Venue  \
0               -79.293031                   Glen Manor Ravine   
1               -79.293031  The Big Carrot Natural Food Market   
2               -79.293031                 Grover Pub and Grub   
3               -79.293031                       Upper Beaches   
4               -79.293031                 Seaspray Restaurant   
5               -79.352188                       MenEssentials   
6               -79.352188                            Pantheon   
7               -79.352188                          La Diperie   
8               -79.352188                        Dolce Gelato   
9               -79.352188                     Cafe Fiorentina   
10              -79.352188                               Mezes   
11              -79.352188              Louis Cifer Brew Works   
12              -79.352188             Messini Authentic Gyros   
13              -79.352188                 Valley Farm Produce   
14              -79.352188                Moksha Yoga Danforth   

    Venue Latitude  Venue Longitude           Venue Category  
0        43.676821       -79.293942                    Trail  
1        43.678879       -79.297734        Health Food Store  
2        43.679181       -79.297215                      Pub  
3        43.680563       -79.292869             Neighborhood  
4        43.678888       -79.298167         Asian Restaurant  
5        43.677820       -79.351265           Cosmetics Shop  
6        43.677621       -79.351434         Greek Restaurant  
7        43.677702       -79.352265           Ice Cream Shop  
8        43.677773       -79.351187           Ice Cream Shop  
9        43.677743       -79.350115       Italian Restaurant  
10       43.677962       -79.350196         Greek Restaurant  
11       43.677663       -79.351313                  Brewery  
12       43.677704       -79.350480         Greek Restaurant  
13       43.677999       -79.349969  Fruit & Vegetable Store  
14       43.677622       -79.352116              Yoga Studio

Now I check how many venues are returned in each neighborhood

In [41]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            58   
Brockton, Parkdale Village, Exhibition Place                           22   
Business reply mail Processing Centre, South Ce...                     19   
CN Tower, King and Spadina, Railway Lands, Harb...                     15   
Central Bay Street                                                     65   
Christie                                                               17   
Church and Wellesley                                                   75   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             35   
Davisville North                                                        8   
Dufferin, Dovercourt Village                                           15   
First Canadian Place, Underground city                                100   
Forest Hill North & West, Forest Hill Road Park                         4   
Garden District, Ryerson                                              100   
Harbourfront East, Union Station, Toronto Islands                     100   
High Park, The Junction South                                          25   
India Bazaar, The Beaches West                                         23   
Kensington Market, Chinatown, Grange Park                              60   
Lawrence Park                                                           3   
Little Portugal, Trinity                                               44   
Moore Park, Summerhill East                                             1   
North Toronto West, Lawrence Park                                      18   
Parkdale, Roncesvalles                                                 15   
Queen's Park, Ontario Provincial Government                            32   
Regent Park, Harbourfront                                              44   
Richmond, Adelaide, King                                               94   
Rosedale                                                                4   
Roselawn                                                                3   
Runnymede, Swansea                                                     35   
St. James Town                                                         79   
St. James Town, Cabbagetown                                            42   
Stn A PO Boxes                                                         97   
Studio District                                                        42   
Summerhill West, Rathnelly, South Hill, Forest ...                     16   
The Annex, North Midtown, Yorkville                                    21   
The Beaches                                                             5   
The Danforth West, Riverdale                                           43   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                         35   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             58   
Brockton, Parkdale Village, Exhibition Place                            22   
Business reply mail Processing Centre, South Ce...                      19   
CN Tower, King and Spadina, Railway Lands, Harb...                      15   
Central Bay Street                                                      65   
Christie                                                                17   
Church and Wellesley                                                    75   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                              35   
Davisville North                    

Let's find how many uniqus categories can be curated from all the return vaues

In [42]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 233 uniques categories.


# 3.2 Analyse  each Neighborhood

In [43]:
# Transform data using the one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Afghan Restaurant  Airport  Airport Food Court  \
0            0                  0        0                   0   
1            0                  0        0                   0   
2            0                  0        0                   0   
3            0                  0        0                   0   
4            0                  0        0                   0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0             0         0            0           0                    0   
1             0         0            0           0                    0   
2             0         0            0           0                    0   
3             0         0            0           0                    0   
4             0         0            0           0                    0   

   Asian Restaurant  Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  \
0                 0                   0              0          0           0   
1                 0                   0              0          0           0   
2                 0                   0              0          0           0   
3                 0                   0              0          0           0   
4                 1                   0              0          0           0   

   Bagel Shop  Bakery  Bank  Bar  Baseball Stadium  Basketball Stadium  Beach  \
0           0       0     0    0                 0                   0      0   
1           0       0     0    0                 0                   0      0   
2           0       0     0    0                 0                   0      0   
3           0       0     0    0                 0                   0      0   
4           0       0     0    0                 0                   0      0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bistro  \
0                0         0           0                   0       0   
1                0         0           0                   0       0   
2                0         0           0                   0       0   
3                0         0           0                   0       0   
4                0         0           0                   0       0   

   Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  \
0              0          0         0                     0               0   
1              0          0         0                     0               0   
2              0          0         0                     0               0   
3              0          0         0                     0               0   
4              0          0         0                     0               0   

   Brewery  Bubble Tea Shop  Building  Burger Joint  Burrito Place  Bus Line  \
0        0                0         0             0              0         0   
1        0                0         0             0              0         0   
2        0                0         0             0              0         0   
3        0                0         0             0              0         0   
4        0                0         0             0              0         0   

   Butcher  Café  Cajun / Creole Restaurant  Camera Store  Candy Store  \
0        0     0                          0             0            0   
1        0     0                          0             0            0   
2        0     0                          0             0            0   
3        0     0                 

In [44]:
toronto_onehot.shape ## Examine the size

(1614, 233)

#### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [45]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Yoga Studio  \
0                                         Berczy Park     0.000000   
1        Brockton, Parkdale Village, Exhibition Place     0.000000   
2   Business reply mail Processing Centre, South C...     0.000000   
3   CN Tower, King and Spadina, Railway Lands, Har...     0.000000   
4                                  Central Bay Street     0.015385   
5                                            Christie     0.000000   
6                                Church and Wellesley     0.026667   
7                      Commerce Court, Victoria Hotel     0.000000   
8                                          Davisville     0.000000   
9                                    Davisville North     0.000000   
10                       Dufferin, Dovercourt Village     0.000000   
11             First Canadian Place, Underground city     0.000000   
12    Forest Hill North & West, Forest Hill Road Park     0.000000   
13                           Garden District, Ryerson     0.000000   
14  Harbourfront East, Union Station, Toronto Islands     0.000000   
15                      High Park, The Junction South     0.000000   
16                     India Bazaar, The Beaches West     0.000000   
17          Kensington Market, Chinatown, Grange Park     0.000000   
18                                      Lawrence Park     0.000000   
19                           Little Portugal, Trinity     0.022727   
20                        Moore Park, Summerhill East     0.000000   
21                  North Toronto West, Lawrence Park     0.055556   
22                             Parkdale, Roncesvalles     0.000000   
23        Queen's Park, Ontario Provincial Government     0.031250   
24                          Regent Park, Harbourfront     0.022727   
25                           Richmond, Adelaide, King     0.000000   
26                                           Rosedale     0.000000   
27                                           Roselawn     0.000000   
28                                 Runnymede, Swansea     0.028571   
29                                     St. James Town     0.000000   
30                        St. James Town, Cabbagetown     0.000000   
31                                     Stn A PO Boxes     0.010309   
32                                    Studio District     0.023810   
33  Summerhill West, Rathnelly, South Hill, Forest...     0.000000   
34                The Annex, North Midtown, Yorkville     0.000000   
35                                        The Beaches     0.000000   
36                       The Danforth West, Riverdale     0.023256   
37           Toronto Dominion Centre, Design Exchange     0.000000   
38                     University of Toronto, Harbord     0.028571   

    Afghan Restaurant   Airport  Airport Food Court  Airport Lounge  \
0            0.000000  0.000000            0.000000        0.000000   
1            0.000000  0.000000            0.000000        0.000000   
2            0.000000  0.000000            0.000000        0.000000   
3            0.000000  0.066667            0.066667        0.133333   
4            0.000000  0.000000            0.000000        0.000000   
5            0.000000  0.000000            0.000000        0.000000   
6            0.013333  0.000000            0.000000        0.000000   
7            0.000000  0.000000            0.000000        0.000000   
8            0.000000  0.000000            0.000000        0.000000   
9            0.000000  0.000000            0.000000        0.000000   
10           0.000000  0.000000            0.000000        0.000000   
11           0.000000  0.000000            0.000000        0.000000   
12           0.000000  0.000000            0.000000        0.000000   
13           0.000000  0.000000            0.000000        0.000000   
14           0.000000  0.000000            0.000000        0.000000   
15           0.000000  0.000000            0.000000        0.000000   
16           0.000000  0.000000  

In [46]:
toronto_grouped.shape

(39, 233)

### Lets print each neighborhood with the top 5 most common venues

In [47]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
                venue  freq
0         Coffee Shop  0.09
1        Cocktail Bar  0.05
2  Seafood Restaurant  0.03
3         Cheese Shop  0.03
4          Restaurant  0.03


----Brockton, Parkdale Village, Exhibition Place----
            venue  freq
0            Café  0.14
1  Breakfast Spot  0.09
2     Coffee Shop  0.09
3       Pet Store  0.05
4    Intersection  0.05


----Business reply mail Processing Centre, South Central Letter Processing Plant Toronto----
                  venue  freq
0    Light Rail Station  0.11
1  Gym / Fitness Center  0.05
2         Auto Workshop  0.05
3            Smoke Shop  0.05
4                  Park  0.05


----CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport----
              venue  freq
0   Airport Service  0.20
1    Airport Lounge  0.13
2       Coffee Shop  0.07
3   Harbor / Marina  0.07
4  Sculpture Garden  0.07


----Central Bay Street----
                 venue  freq
0      

I insert the data into a pandas dataframe

First I write a function to sort the venues in descending order.

In [48]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

###  Lets create a new dataframe and display the top 10 venues of each neighborhood

In [49]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head(15)

Neighborhood 1st Most Common Venue  \
0                                         Berczy Park           Coffee Shop   
1        Brockton, Parkdale Village, Exhibition Place                  Café   
2   Business reply mail Processing Centre, South C...    Light Rail Station   
3   CN Tower, King and Spadina, Railway Lands, Har...       Airport Service   
4                                  Central Bay Street           Coffee Shop   
5                                            Christie         Grocery Store   
6                                Church and Wellesley           Coffee Shop   
7                      Commerce Court, Victoria Hotel           Coffee Shop   
8                                          Davisville           Pizza Place   
9                                    Davisville North                   Gym   
10                       Dufferin, Dovercourt Village              Pharmacy   
11             First Canadian Place, Underground city           Coffee Shop   
12    Forest Hill North & West, Forest Hill Road Park         Jewelry Store   
13                           Garden District, Ryerson        Clothing Store   
14  Harbourfront East, Union Station, Toronto Islands           Coffee Shop   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0           Cocktail Bar    Seafood Restaurant                  Café   
1            Coffee Shop        Breakfast Spot                   Gym   
2                   Park         Auto Workshop            Comic Shop   
3         Airport Lounge      Sculpture Garden               Airport   
4     Italian Restaurant        Sandwich Place   Japanese Restaurant   
5                   Café                  Park    Athletics & Sports   
6       Sushi Restaurant   Japanese Restaurant               Gay Bar   
7             Restaurant                  Café                 Hotel   
8         Sandwich Place          Dessert Shop                   Gym   
9         Sandwich Place                  Park  Gym / Fitness Center   
10                Bakery           Supermarket                  Café   
11                  Café                 Hotel            Restaurant   
12                 Trail    Mexican Restaurant      Sushi Restaurant   
13           Coffee Shop       Bubble Tea Shop                  Café   
14              Aquarium                  Café                 Hotel   

        5th Most Common Venue 6th Most Common Venue  7th Most Common Venue  \
0                      Bakery              Beer Bar            Cheese Shop   
1                      Bakery     Convenience Store  Performing Arts Venue   
2                 Pizza Place      Recording Studio                Butcher   
3          Airport Food Court      Airport Terminal        Harbor / Marina   
4                        Café           Salad Place        Thai Restaurant   
5          Italian Restaurant                 Diner             Restaurant   
6                  Restaurant            Smoke Shop            Men's Store   
7                         Gym   American Restaurant     Seafood Restaurant   
8                        Café    Italian Restaurant            Coffee Shop   
9              Breakfast Spot                 Hotel       Department Store   
10              Grocery Store               Brewery                   Bank   
11                        Gym   American Restaurant    Japanese Restaurant   
12              Women's Store      Department Store   Ethiopian Restaurant   
13  Middle Eastern Restaurant   Japanese Restaurant         Cosmetics Shop   
14         Italian Restaurant               Brewery             Restaurant   

       8th Most Common Venue        9th Most Common Venue  \
0                 Restaurant           Basketball Stadium   
1                  Pet Store                 Climbing Gym   
2                 Restaurant                Burrito Place   
3                   Boutique                          Bar   
4           Department Store                          Bar   
5       

In [50]:
neighborhoods_venues_sorted.shape

(39, 11)

# 3.3 Cluster of the Neighborhoods with K-means clustering algorithm

## 3.3.1 Run K-means to cluster the neighborhoods into 5 clusters and visualise them
its an unsupervised Machine Learning algorithm that help grouping the data in similar datapoints.

In [51]:
#For convenience I rename Toronto_T into df_T and I will add there the Cluster Labels
dfT= Toronto_T

In [52]:
# set numbers of clusters
k=5
toronto_clustering = dfT.drop(['Postal Code','Borough','Neighborhood'],1) # Cluster the data of the bouroughs that contain Toronto in theit name

# run k-means clustering
kmeans = KMeans(n_clusters = k,random_state=0).fit(toronto_clustering)
#Check clusters labels
kmeans.labels_



array([0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4,
       1, 1, 3, 3, 3, 4, 4, 4, 3, 2, 3, 3, 2, 2, 2, 4, 0], dtype=int32)

At the data set dfT I will add the Cluster Labels

In [53]:
dfT.insert(0, 'Cluster Labels', kmeans.labels_) # insert the column Cluster Labels with the values of kmeans.labels_
dfT.head()

Cluster Labels Postal Code          Borough  \
0               0         M4E     East Toronto   
1               0         M4K     East Toronto   
2               0         M4L     East Toronto   
3               0         M4M     East Toronto   
4               1         M4N  Central Toronto   

                     Neighborhood   Latitude  Longitude  
0                     The Beaches  43.676357 -79.293031  
1    The Danforth West, Riverdale  43.679557 -79.352188  
2  India Bazaar, The Beaches West  43.668999 -79.315572  
3                 Studio District  43.659526 -79.340923  
4                   Lawrence Park  43.728020 -79.388790

## 3.3.2  Visualise the clustering at neighbourhoods of the Boroughs that contain the name Toronto 

#### Note for the reviewer: I am not sure if you could see the map in the GitHub. If you run the code the map appears

In [54]:
# create map
map_clusters = folium.Map(location=[43.6534817,-79.3839347],zoom_start=10)

# set color scheme for the clusters
x = np.arange(k)
ys = [i + x + (i*x)**2 for i in range(k)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, neighborhood, cluster in zip(dfT['Latitude'], dfT['Longitude'], dfT['Neighborhood'], dfT['Cluster Labels']):
    label = folium.Popup(' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## 3.3.4 Examine Clusters 

Create a dataframe that includes the cluster as well as the top 5 venues of each neighborhood.

In [55]:
# add clustering labels to a new Data called toronto_merged

toronto_merged = dfT.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head(20) # check the last columns!



Cluster Labels Postal Code           Borough  \
0                0         M4E      East Toronto   
1                0         M4K      East Toronto   
2                0         M4L      East Toronto   
3                0         M4M      East Toronto   
4                1         M4N   Central Toronto   
5                1         M4P   Central Toronto   
6                1         M4R   Central Toronto   
7                1         M4S   Central Toronto   
8                1         M4T   Central Toronto   
9                1         M4V   Central Toronto   
10               4         M4W  Downtown Toronto   
11               4         M4X  Downtown Toronto   
12               4         M4Y  Downtown Toronto   
13               4         M5A  Downtown Toronto   
14               4         M5B  Downtown Toronto   
15               4         M5C  Downtown Toronto   
16               4         M5E  Downtown Toronto   
17               4         M5G  Downtown Toronto   
18               4         M5H  Downtown Toronto   
19               4         M5J  Downtown Toronto   

                                         Neighborhood   Latitude  Longitude  \
0                                         The Beaches  43.676357 -79.293031   
1                        The Danforth West, Riverdale  43.679557 -79.352188   
2                      India Bazaar, The Beaches West  43.668999 -79.315572   
3                                     Studio District  43.659526 -79.340923   
4                                       Lawrence Park  43.728020 -79.388790   
5                                    Davisville North  43.712751 -79.390197   
6                   North Toronto West, Lawrence Park  43.715383 -79.405678   
7                                          Davisville  43.704324 -79.388790   
8                         Moore Park, Summerhill East  43.689574 -79.383160   
9   Summerhill West, Rathnelly, South Hill, Forest...  43.686412 -79.400049   
10                                           Rosedale  43.679563 -79.377529   
11                        St. James Town, Cabbagetown  43.667967 -79.367675   
12                               Church and Wellesley  43.665860 -79.383160   
13                          Regent Park, Harbourfront  43.654260 -79.360636   
14                           Garden District, Ryerson  43.657162 -79.378937   
15                                     St. James Town  43.651494 -79.375418   
16                                        Berczy Park  43.644771 -79.373306   
17                                 Central Bay Street  43.657952 -79.387383   
18                           Richmond, Adelaide, King  43.650571 -79.384568   
19  Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0       Asian Restaurant     Health Food Store                   Pub   
1       Greek Restaurant           Coffee Shop    Italian Restaurant   
2         Sandwich Place                  Park  Fast Food Restaurant   
3                   Café           Coffee Shop                Bakery   
4                   Park           Swim School              Bus Line   
5                    Gym        Sandwich Place                  Park   
6            Coffee Shop        Clothing Store           Yoga Studio   
7            Pizza Place        Sandwich Place          Dessert Shop   
8                   Park         Women's Store         Deli / Bodega   
9                    Pub           Coffee Shop            Bagel Shop   
10                  Park            Playground                 Trail   
11                Bakery           Coffee Shop           Pizza Place   
12           Coffee Shop      Sushi Restaurant   Japanese Restaurant   
13           Coffee Shop                  Park                Bakery   
14        Clothing Store           Coffee Shop       Bubble Tea Shop   
15                  Café           Coffee Shop             Gastropub   
16           Coffee Shop   

In [57]:
toronto_merged.drop(['Latitude','Longitude'], axis=1) # removethe columns of the coordinates

Cluster Labels Postal Code           Borough  \
0                0         M4E      East Toronto   
1                0         M4K      East Toronto   
2                0         M4L      East Toronto   
3                0         M4M      East Toronto   
4                1         M4N   Central Toronto   
5                1         M4P   Central Toronto   
6                1         M4R   Central Toronto   
7                1         M4S   Central Toronto   
8                1         M4T   Central Toronto   
9                1         M4V   Central Toronto   
10               4         M4W  Downtown Toronto   
11               4         M4X  Downtown Toronto   
12               4         M4Y  Downtown Toronto   
13               4         M5A  Downtown Toronto   
14               4         M5B  Downtown Toronto   
15               4         M5C  Downtown Toronto   
16               4         M5E  Downtown Toronto   
17               4         M5G  Downtown Toronto   
18               4         M5H  Downtown Toronto   
19               4         M5J  Downtown Toronto   
20               4         M5K  Downtown Toronto   
21               4         M5L  Downtown Toronto   
22               1         M5N   Central Toronto   
23               1         M5P   Central Toronto   
24               3         M5R   Central Toronto   
25               3         M5S  Downtown Toronto   
26               3         M5T  Downtown Toronto   
27               4         M5V  Downtown Toronto   
28               4         M5W  Downtown Toronto   
29               4         M5X  Downtown Toronto   
30               3         M6G  Downtown Toronto   
31               2         M6H      West Toronto   
32               3         M6J      West Toronto   
33               3         M6K      West Toronto   
34               2         M6P      West Toronto   
35               2         M6R      West Toronto   
36               2         M6S      West Toronto   
37               4         M7A  Downtown Toronto   
38               0         M7Y      East Toronto   

                                         Neighborhood 1st Most Common Venue  \
0                                         The Beaches      Asian Restaurant   
1                        The Danforth West, Riverdale      Greek Restaurant   
2                      India Bazaar, The Beaches West        Sandwich Place   
3                                     Studio District                  Café   
4                                       Lawrence Park                  Park   
5                                    Davisville North                   Gym   
6                   North Toronto West, Lawrence Park           Coffee Shop   
7                                          Davisville           Pizza Place   
8                         Moore Park, Summerhill East                  Park   
9   Summerhill West, Rathnelly, South Hill, Forest...                   Pub   
10                                           Rosedale                  Park   
11                        St. James Town, Cabbagetown                Bakery   
12                               Church and Wellesley           Coffee Shop   
13                          Regent Park, Harbourfront           Coffee Shop   
14                           Garden District, Ryerson        Clothing Store   
15                                     St. James Town                  Café   
16                                        Berczy Park           Coffee Shop   
17                                 Central Bay Street           Coffee Shop   
18                           Richmond, Adelaide, King           Coffee Shop   
19  Harbourfront East, Union Station, Toronto Islands           Coffee Shop   
20           Toronto Dominion Centre, Design Exchange           Coffee Shop   
21                     Commerce Court, Victoria Hotel           Coffee Shop   
22                                           Roselawn                  Pool   
23    Forest Hill North & W

Now, we examine each cluster and determine the discriminating venue categories that distinguish each cluster. Based on the defining categories, we can then assign a name to each cluster. 

### Cluster 1

In [62]:
#I've just select the first cluster with the following line that worked for me
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0]

## The following line didnt work!!
 ##toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]] 

Cluster Labels Postal Code       Borough  \
0                0         M4E  East Toronto   
1                0         M4K  East Toronto   
2                0         M4L  East Toronto   
3                0         M4M  East Toronto   
38               0         M7Y  East Toronto   

                                         Neighborhood   Latitude  Longitude  \
0                                         The Beaches  43.676357 -79.293031   
1                        The Danforth West, Riverdale  43.679557 -79.352188   
2                      India Bazaar, The Beaches West  43.668999 -79.315572   
3                                     Studio District  43.659526 -79.340923   
38  Business reply mail Processing Centre, South C...  43.662744 -79.321558   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0       Asian Restaurant     Health Food Store                   Pub   
1       Greek Restaurant           Coffee Shop    Italian Restaurant   
2         Sandwich Place                  Park  Fast Food Restaurant   
3                   Café           Coffee Shop                Bakery   
38    Light Rail Station                  Park         Auto Workshop   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0                  Trail   Distribution Center      Department Store   
1              Bookstore        Ice Cream Shop            Restaurant   
2            Pizza Place                   Gym     Food & Drink Shop   
3              Gastropub   American Restaurant               Brewery   
38            Comic Shop           Pizza Place      Recording Studio   

     7th Most Common Venue      8th Most Common Venue 9th Most Common Venue  \
0             Dessert Shop                      Diner        Discount Store   
1   Furniture / Home Store               Liquor Store                   Spa   
2             Liquor Store              Burrito Place            Restaurant   
3              Yoga Studio  Latin American Restaurant                  Food   
38                 Butcher                 Restaurant         Burrito Place   

   10th Most Common Venue  
0        Doner Restaurant  
1     Japanese Restaurant  
2      Italian Restaurant  
3          Ice Cream Shop  
38                Brewery

### Cluster 2

In [63]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1]

Cluster Labels Postal Code          Borough  \
4                1         M4N  Central Toronto   
5                1         M4P  Central Toronto   
6                1         M4R  Central Toronto   
7                1         M4S  Central Toronto   
8                1         M4T  Central Toronto   
9                1         M4V  Central Toronto   
22               1         M5N  Central Toronto   
23               1         M5P  Central Toronto   

                                         Neighborhood   Latitude  Longitude  \
4                                       Lawrence Park  43.728020 -79.388790   
5                                    Davisville North  43.712751 -79.390197   
6                   North Toronto West, Lawrence Park  43.715383 -79.405678   
7                                          Davisville  43.704324 -79.388790   
8                         Moore Park, Summerhill East  43.689574 -79.383160   
9   Summerhill West, Rathnelly, South Hill, Forest...  43.686412 -79.400049   
22                                           Roselawn  43.711695 -79.416936   
23    Forest Hill North & West, Forest Hill Road Park  43.696948 -79.411307   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
4                   Park           Swim School              Bus Line   
5                    Gym        Sandwich Place                  Park   
6            Coffee Shop        Clothing Store           Yoga Studio   
7            Pizza Place        Sandwich Place          Dessert Shop   
8                   Park         Women's Store         Deli / Bodega   
9                    Pub           Coffee Shop            Bagel Shop   
22                  Pool          Home Service                Garden   
23         Jewelry Store                 Trail    Mexican Restaurant   

   4th Most Common Venue    5th Most Common Venue 6th Most Common Venue  \
4     Falafel Restaurant              Event Space  Ethiopian Restaurant   
5   Gym / Fitness Center           Breakfast Spot                 Hotel   
6    Sporting Goods Shop  Health & Beauty Service    Italian Restaurant   
7                    Gym                     Café    Italian Restaurant   
8            Event Space     Ethiopian Restaurant     Electronics Store   
9       Sushi Restaurant                     Bank            Sports Bar   
22         Women's Store            Deli / Bodega  Ethiopian Restaurant   
23      Sushi Restaurant            Women's Store      Department Store   

          7th Most Common Venue        8th Most Common Venue  \
4             Electronics Store  Eastern European Restaurant   
5              Department Store            Food & Drink Shop   
6          Fast Food Restaurant                        Diner   
7                   Coffee Shop             Sushi Restaurant   
8   Eastern European Restaurant                   Donut Shop   
9           Fried Chicken Joint                   Restaurant   
22            Electronics Store  Eastern European Restaurant   
23         Ethiopian Restaurant            Electronics Store   

          9th Most Common Venue 10th Most Common Venue  
4                    Donut Shop       Doner Restaurant  
5                         Diner         Discount Store  
6            Mexican Restaurant     Chinese Restaurant  
7                          Park       Indoor Play Area  
8              Doner Restaurant                Dog Run  
9                   Pizza Place            Supermarket  
22                   Donut Shop       Doner Restaurant  
23  Eastern European Restaurant             Donut Shop

### Cluster 3

In [64]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2]

Cluster Labels Postal Code       Borough                   Neighborhood  \
31               2         M6H  West Toronto   Dufferin, Dovercourt Village   
34               2         M6P  West Toronto  High Park, The Junction South   
35               2         M6R  West Toronto         Parkdale, Roncesvalles   
36               2         M6S  West Toronto             Runnymede, Swansea   

     Latitude  Longitude 1st Most Common Venue 2nd Most Common Venue  \
31  43.669005 -79.442259              Pharmacy                Bakery   
34  43.661608 -79.464763         Grocery Store                  Café   
35  43.648960 -79.456325        Breakfast Spot             Gift Shop   
36  43.651571 -79.484450                  Café                   Pub   

   3rd Most Common Venue 4th Most Common Venue        5th Most Common Venue  \
31           Supermarket                  Café                Grocery Store   
34    Mexican Restaurant       Thai Restaurant                        Diner   
35          Dessert Shop         Movie Theater  Eastern European Restaurant   
36           Pizza Place           Coffee Shop             Sushi Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
31               Brewery                  Bank                  Park   
34  Fast Food Restaurant    Italian Restaurant   Fried Chicken Joint   
35               Dog Run    Italian Restaurant                   Bar   
36    Italian Restaurant                   Gym          Gourmet Shop   

        9th Most Common Venue     10th Most Common Venue  
31                Music Venue                        Bar  
34  Cajun / Creole Restaurant     Furniture / Home Store  
35                       Bank                 Restaurant  
36                 Restaurant  Latin American Restaurant

### Cluster 4

In [65]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3]

Cluster Labels Postal Code           Borough  \
24               3         M5R   Central Toronto   
25               3         M5S  Downtown Toronto   
26               3         M5T  Downtown Toronto   
30               3         M6G  Downtown Toronto   
32               3         M6J      West Toronto   
33               3         M6K      West Toronto   

                                    Neighborhood   Latitude  Longitude  \
24           The Annex, North Midtown, Yorkville  43.672710 -79.405678   
25                University of Toronto, Harbord  43.662696 -79.400049   
26     Kensington Market, Chinatown, Grange Park  43.653206 -79.400049   
30                                      Christie  43.669542 -79.422564   
32                      Little Portugal, Trinity  43.647927 -79.419750   
33  Brockton, Parkdale Village, Exhibition Place  43.636847 -79.428191   

   1st Most Common Venue  2nd Most Common Venue  3rd Most Common Venue  \
24        Sandwich Place                   Café            Coffee Shop   
25                  Café              Bookstore                    Bar   
26                  Café  Vietnamese Restaurant           Dessert Shop   
30         Grocery Store                   Café                   Park   
32                   Bar             Restaurant  Vietnamese Restaurant   
33                  Café            Coffee Shop         Breakfast Spot   

            4th Most Common Venue 5th Most Common Venue  \
24                           Park           Pizza Place   
25             Italian Restaurant   Japanese Restaurant   
26  Vegetarian / Vegan Restaurant                Bakery   
30             Athletics & Sports    Italian Restaurant   
32  Vegetarian / Vegan Restaurant                  Café   
33                            Gym                Bakery   

        6th Most Common Venue  7th Most Common Venue 8th Most Common Venue  \
24  Middle Eastern Restaurant                    Pub             BBQ Joint   
25                 Restaurant                 Bakery              Beer Bar   
26         Mexican Restaurant            Coffee Shop         Grocery Store   
30                      Diner             Restaurant            Baby Store   
32                Men's Store       Asian Restaurant           Yoga Studio   
33          Convenience Store  Performing Arts Venue             Pet Store   

   9th Most Common Venue 10th Most Common Venue  
24            Donut Shop         History Museum  
25            Beer Store         Sandwich Place  
26                   Bar            Pizza Place  
30           Candy Store              Nightclub  
32          Cupcake Shop            Record Shop  
33          Climbing Gym             Restaurant

### Cluster 5

In [66]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4]

Cluster Labels Postal Code           Borough  \
10               4         M4W  Downtown Toronto   
11               4         M4X  Downtown Toronto   
12               4         M4Y  Downtown Toronto   
13               4         M5A  Downtown Toronto   
14               4         M5B  Downtown Toronto   
15               4         M5C  Downtown Toronto   
16               4         M5E  Downtown Toronto   
17               4         M5G  Downtown Toronto   
18               4         M5H  Downtown Toronto   
19               4         M5J  Downtown Toronto   
20               4         M5K  Downtown Toronto   
21               4         M5L  Downtown Toronto   
27               4         M5V  Downtown Toronto   
28               4         M5W  Downtown Toronto   
29               4         M5X  Downtown Toronto   
37               4         M7A  Downtown Toronto   

                                         Neighborhood   Latitude  Longitude  \
10                                           Rosedale  43.679563 -79.377529   
11                        St. James Town, Cabbagetown  43.667967 -79.367675   
12                               Church and Wellesley  43.665860 -79.383160   
13                          Regent Park, Harbourfront  43.654260 -79.360636   
14                           Garden District, Ryerson  43.657162 -79.378937   
15                                     St. James Town  43.651494 -79.375418   
16                                        Berczy Park  43.644771 -79.373306   
17                                 Central Bay Street  43.657952 -79.387383   
18                           Richmond, Adelaide, King  43.650571 -79.384568   
19  Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752   
20           Toronto Dominion Centre, Design Exchange  43.647177 -79.381576   
21                     Commerce Court, Victoria Hotel  43.648198 -79.379817   
27  CN Tower, King and Spadina, Railway Lands, Har...  43.628947 -79.394420   
28                                     Stn A PO Boxes  43.646435 -79.374846   
29             First Canadian Place, Underground city  43.648429 -79.382280   
37        Queen's Park, Ontario Provincial Government  43.662301 -79.389494   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
10                  Park            Playground                 Trail   
11                Bakery           Coffee Shop           Pizza Place   
12           Coffee Shop      Sushi Restaurant   Japanese Restaurant   
13           Coffee Shop                  Park                Bakery   
14        Clothing Store           Coffee Shop       Bubble Tea Shop   
15                  Café           Coffee Shop             Gastropub   
16           Coffee Shop          Cocktail Bar    Seafood Restaurant   
17           Coffee Shop    Italian Restaurant        Sandwich Place   
18           Coffee Shop            Restaurant                  Café   
19           Coffee Shop              Aquarium                  Café   
20           Coffee Shop                 Hotel                  Café   
21           Coffee Shop            Restaurant                  Café   
27       Airport Service        Airport Lounge      Sculpture Garden   
28           Coffee Shop                  Café    Seafood Restaurant   
29           Coffee Shop                  Café                 Hotel   
37           Coffee Shop                 Diner      Sushi Restaurant   

   4th Most Common Venue      5th Most Common Venue 6th Most Common Venue  \
10          Dance Studio       Ethiopian Restaurant     Electronics Store   
11    Italian Restaurant                       Café                   Pub   
12               Gay Bar                 Restaurant            Smoke Shop   
13                   Pub             Breakfast Spot                  Café   
14                  Café  Middle Eastern Restaurant   Japanese Restaurant   
15            Restaurant               Cocktail Bar   American Restaurant   
16         

This is the  Notebook that contains all the assigment of the 3rd week of the capstone project.
This is the part 3 which is the continuation of Part 1  and part 2 with the added task perform the Exploration, Analysis and Clustering of a boroughs and clustering using the kmeans method.  

Thank you for reviewing this Notebook.

Best wishes,

Mary 